### ISSUE TITLE
Is there a way to output loss value for every iteration

### ISSUE THREAD

**Issue by scrcro** [mlpack/mlpack#1455]
If I need to understand how the loss value behaves with every iteration of the training, is there a way to output the loss value to a file? I know there is some output in the form of LOG::Info but I just need the raw values without the texts. Is there a good way to accomplish this?

**Comment by rcurtin** [comment]
Hmm, there is not an easy way to do it (and maybe there should be).  Right now the three ways I can think of are:

 * Capture the `Log::Info` output that you already get during training and process it a little bit to recover the raw values only with awk and sed (or whatever tool).

 * Modify the SGD optimizer output in `src/mlpack/core/optimizers/sgd/` to get what you need and print it to `std::cout`.

 * Run the training in a single-epoch loop and print what you need, i.e.,

```
// I'm assuming opt is whatever optimizer you're using.
opt.MaxIterations() = data.n_cols; // so it will take one pass over the data
for (size_t i = 0; i < epochs; ++i)
{
  model.Train(trainingData, trainingResponses, opt);
  std::cout << CalculateLoss(model); // for some CalculateLoss() function
}
```

If you have ideas on how we could improve the overall abstraction to do things like this, I certainly have no opposition to that.  Maybe a template parameter to optimizers to perform some behavior at the end of each epoch or something?

**Comment by zoq** [comment]
Another idea would be to forward the function calls, something like:

```c++
template<class FunctionType>
class WrapperFunction
{
 public:
  WrapperFunction(FunctionType& function) : function(function) { }

  arma::mat& Coordinates() { return coordinates; }

  void Shuffle() { function.Shuffle(); }

  size_t NumFunctions() const { return function.NumFunctions(); }

  arma::mat GetInitialPoint() const { return function.GetInitialPoint(); }

  double Evaluate(const arma::mat& coordinates,
                  const size_t begin,
                  const size_t batchSize)
  {
    const double result = function.Evaluate(coordinates, begin, batchSize);

    // Save the output here.

    return result;
  }

  double Evaluate(const arma::mat& coordinates)
  {
    const double result = function.Evaluate(coordinates);

    // Save the output here.

    return result;
  }

  void Gradient(const arma::mat& coordinates,
                const size_t begin,
                arma::mat& gradient,
                const size_t batchSize)
  {
    function.Gradient(coordinates, begin, gradient, batchSize);
  }

  void Gradient(const arma::mat& coordinates, arma::mat& gradient)
  {
    function.Gradient(coordinates, gradient);
  }

 private:
  FunctionType& function;
  arma::mat coordinates;
};
```

**Comment by scrcro** [comment]
Thank you @rcurtin @zoq. Let me take a look at the optimizers' infrastructure. 

@zoq Are you suggesting that we have additional methods for say FFN or RNN?

**Comment by zoq** [comment]
The idea builds on top of almost any mlpack method, that allows you to intercept the training process, but if you can modify the existing codebase you could directly modify the necessary parts as well. Here is a simple example, first the standard way:

```c++
FFN<NegativeLogLikelihood<>, RandomInitialization, CustomLayer<> > model;
model.Add<Linear<> >(trainData.n_rows, 8);
model.Add<CustomLayer<> >();
model.Add<Linear<> >(8, 3);
model.Add<LogSoftMax<> >();

RMSProp opt(0.01, 32, 0.88, 1e-8, 15, -1);
model.Train(trainData, trainLabels, opt);
```

before we can use the idea from above we have to make ``void ResetData(arma::mat predictors, arma::mat responses);`` public.

```c++
FFN<NegativeLogLikelihood<>, RandomInitialization, CustomLayer<> > model;
model.Add<Linear<> >(trainData.n_rows, 8);
model.Add<CustomLayer<> >();
model.Add<Linear<> >(8, 3);
model.Add<LogSoftMax<> >();

WrapperFunction<decltype(model)> modelFunction(model);

RMSProp opt(0.01, 32, 0.88, 1e-8, 15, -1);
opt.Optimize(modelFunction, model.Parameters());
```

inside the ``Evaluate`` function of the ``WrapperFunction`` class you can save the loss or whatever you like to do. I hope this was helpful, let me know if I should clarify anything.

**Comment by rcurtin** [comment]
Right, that idea allows no changes at all to the existing codebase, which is nice.  But I wonder if it would be nicer for users in the long term to allow something a little more like Keras does, by making any extra functionality a part of the optimizer:

```
// I'm just using names of Keras callbacks here plus ones I make up...

// Here's one way to do it.
RMSProp<EarlyStopping, PrintLoss> opt(...);

// Or maybe it can be passed to the Optimize() call.
opt.Optimize(model, model.Parameters(), std::make_tuple(EarlyStopping(), PrintLoss()));

// Actually I guess more typically these are trained by calling model.Train(), so maybe we can
// pass them there instead?
model.Train(trainData, trainResponses, opt, std::make_tuple(EarlyStopping(), PrintLoss()));
// A little bit of template metaprogramming above should make it easy to differentiate between
// optimizers and other behavior.
```

What do you think?  It would be easy to add this support to the `FNN` and `RNN` classes, but harder to add it to all the optimizers (though more generic and not specific to neural networks---which may or may not be a good idea, I don't know).

**Comment by zoq** [comment]
> What do you think? It would be easy to add this support to the FNN and RNN classes, but harder to add it to all the optimizers (though more generic and not specific to neural networks---which may or may not be a good idea, I don't know).

Having some kind of callback functionality e.g. to print the loss is a good idea for almost all methods including NN's, so I would opt for a optimizer level solution. The issue I see is what if we like to print/store/process additional optimizer information like the gradient, ideally, someone could pass a generic function and can handle everything inside that function e.g. printing the loss and gradient. Don't see an easy solution for the idea, since there is no easy way to access optimizer specific parameter, so something like

```c++
Optimize(...)
{
  PrintLoss(this);
}

template<class OptimizerType>
PrintLoss(OptimizerType& optimizer)
{
  std::cout << optimizer.Loss() << std::endl;
  std::cout << optimizer.Gradient() << std::endl;
}
```

would be great to have but this would mean that each optimizer has to expose internal parameter. Another solution is to pass every relevant parameter inside the Optimizer function itself, something like:

```c++
Optimize(...)
{
  PrintLoss(loss, gradient, iterate);
}

template<>
template<typename... Targs>
void Update(const double loss, const arma::mat& gradient, Targs... fargs)
{
  std::cout << loss << std::endl;
  std::cout << gradient << std::endl;
}
```

what I don't like about the solution is that the order is fixed.

**Comment by scrcro** [comment]
Thank you @zoq  and @rcurtin . I need a quick solution now so perhaps I will use the wrapper technique. Though I need to study the optimizers closely to see if they can actually return the values for every iteration.
For a long term generic solution we perhaps need some sort of call back  mechanism like Keras.

**Comment by rcurtin** [comment]
Ah, I agree, but maybe we can combine both ideas to keep the syntax nice(ish):

```
opt.Optimize(function, parameters, std::make_tuple(EarlyStopping(), PrintLoss()));
```

and then:

```
// syntax is probably not perfect here, I am writing as I go without checking :)
template<typename FunctionType, typename... Targs>
void Optimize(FunctionType& f, arma::mat parameters, std::tuple<Targs...> t)
{
  FunctionWrapper<FunctionType, Targs...> fw(f, t);
}
```

and then `FunctionWrapper::Evaluate()` is something that calls `Evaluate()` on the first item in the tuple, which then passes it to the next:

```
class PrintLoss
{
  // I'm almost certain this doesn't work quite right but the idea is there...
  // I guess we have to check for the maximum value of Index.  Maybe it can be simplified.
  template<size_t Index, typename... Targs>
  double Evaluate(const arma::mat& parameters, std::tuple<Targs...>& t)
  {
    const double loss = std::get<Index>(t).Evaluate<Index + 1, Targs...>(parameters, t);
    std::cout << "Hi there, the loss is " << loss << "!" << std::endl;
  }
};
```

For accessing optimizer parameters, etc., I wonder if some combination of a traits class and public accessors could be used.

What do you think?  I'm happy with the user interface of just passing instantiated classes to the `Optimize()` call (and maybe this could be wrapped for some methods as part of the `Train()` call), so if you think the basic idea of what I've sketched here could work, then maybe it would be a good way to go.

**Comment by zoq** [comment]
I like the idea, all I like to see is a really minimal class without any boilerplate code just to make it work, so the optimization step is basically all I have to implement. I think what we have right now is exactly that, the recently added ``DecomposableFunctionTypeAPI`` feature makes it even simpler.

**Comment by rcurtin** [comment]
> so the optimization step is basically all I have to implement

I'm not sure I fully understand, but I want to be sure before I go implement a prototype---do you mean that you'd like to see it easy for a user to add different functionality like printing the loss, early stopping, etc. to an optimizer?  Or do you mean easy for a maintainer to add some kind of new hook?

I'm happy to try and make a prototype for it, but I want to make sure that I'll make something we can all be happy with. :)

**Comment by zoq** [comment]
Actually in this case, I was talking about understanding and adding a new optimizer. Right now, the only requirement is to implement ``Optimize(...)``, there is no need to add all kinds of accesors etc. And if I take the Adam code as an example:

```
 // Increment the iteration counter variable.
++iteration;

// And update the iterate.
m *= beta1;
m += (1 - beta1) * gradient;

v *= beta2;
v += (1 - beta2) * (gradient % gradient);

const double biasCorrection1 = 1.0 - std::pow(beta1, iteration);
const double biasCorrection2 = 1.0 - std::pow(beta2, iteration);
```

it directly matches with the equations from the paper. I think the idea above is a really neat way to transparently add the features we need, at the same time, there is no need to implement them.

**Comment by scrcro** [comment]
@zoq  I used the wrapper technique as suggested above. I'm storing the result of the Evaluate call as the iteration loss. But the loss seems to be oscillating around a fairly high value. Is the result of the Evaluate call the right loss value?
I tried to use the same data on a similar model using Keras. This gives me the correct chracteristics. 
I am using RNN with LSTM.

**Comment by zoq** [comment]
Sounds like the stepsize is to high or some issue with the weight distribution. Does it get better after a few iterations, also do you think you could share a minimal example to reproduce the issue?

**Comment by scrcro** [comment]
I tried with a relatively small step size (0.001) as well. And with a batch size of 1 as well. I will post the code soon. I tried both RMSProp and Adam optimizer. I also have a custom layer. Not sure if that is the problem. At the moment I want to be sure that I am capturing the correct loss value. Is the loss cumulative on the batch? I mean will it add up every loss within the batch?

**Comment by zoq** [comment]
I guess, you see similar results with ``stepsize = 0.001`` and ``batchSize = 1``; and you are correct the loss is cumulative for the current batch.

**Comment by scrcro** [comment]
Yes, the absolute loss value is less but the oscillation characteristics remain, whether step size is 0.05 or 0.001. Is the result of `f.EvaluateWithGradient` the correct value of the loss?

**Comment by scrcro** [comment]
Just occurred to me. Will the cumulative loss cause the parameters to overadjust? Could that be the cause of the oscillations?

**Comment by zoq** [comment]
> Is the result of f.EvaluateWithGradient the correct value of the loss?
Yes, EvaluateWithGradient works.

> Will the cumulative loss cause the parameters to overadjust?
If the batch size is constant, the tolerance will be in the same range, so not sure I see the effect; perhaps I missed something?

Also, we should keep in mind that this has the effect the stepsize should be selected according to the batch size.

**Comment by scrcro** [comment]
The parameter adjustment is eventually a factor of the loss, numerically.
So if the loss absolute value is high the parameters may be overadjusted.
I'm speculating here really. I have not looked at the maths. Perhaps same
effect as the high step size?

**Comment by zoq** [comment]
Do you use MSE as your loss function? If that is the case can you make sure you the Forward function looks like:

https://github.com/mlpack/mlpack/blob/3b7bbf0f14172cdb00fd16cbf12918b07c888b96/src/mlpack/methods/ann/loss_functions/mean_squared_error_impl.hpp#L27-L33

**Comment by scrcro** [comment]
No it is multi class problem . I'm using NegativeLogLikelihood.

**Comment by zoq** [comment]
Okay, thanks for the fast response, I'll see if I can reproduce the issue you are seeing in the next days, perhaps I can use one of the existing tests, can you tell me more about your model architecture?

**Comment by scrcro** [comment]
Yes sure. Here is the model:
EmbeddingLayer (custom)
-->LinearNoBias-->Sigmoid-->LSTM-->LinearNoBias-->LogSoftMax-->NegativeLogLikelihood.

EmbeddingLayer essentially takes a set of intergers, signifying a sequence
of words, as input and outputs a vector = dimension*no of words.
LinearNoBias scales the vector down to number of LSTM units, which I use as
a hyper parameter. Then LSTM to output through a linear layer since the
output = vocab size which depends on the data.

network
      .template Add< EmbeddingLayer< UpdatePolicyType, InitializationType >
>(
          vocabSize, dim, 1, stepSize, updatePolicy, initializer);
  network.template Add< mlpack::ann::LinearNoBias<> >(dim, hiddenUnits);
  network.template Add< mlpack::ann::SigmoidLayer<> >();
  network.template Add< mlpack::ann::LSTM<> >(hiddenUnits, hiddenUnits);
  network.template Add< mlpack::ann::LinearNoBias<> >(hiddenUnits,
vocabSize);
  network.template Add< mlpack::ann::LogSoftMax<> >();

**Comment by zoq** [comment]
I was able to reproduce the issue using the ``ReberGrammar`` test, it looks like the stepsize is too high, also it does not train on the full training set at each iteration, which might interfere with the batch setup. As a simple solution, I switched the optimizer to ``RMSProp`` (``RMSProp opt(0.005, 100, 0.9, 1e-08, 50000, 1e-5);``) which adapts the stepsize. The setup workes a lot better, perhaps you can observe the same for your model? Also do you use a similair setup as for the ``ReberGrammar`` soemthing like:


https://github.com/mlpack/mlpack/blob/3b7bbf0f14172cdb00fd16cbf12918b07c888b96/src/mlpack/tests/recurrent_network_test.cpp#L510-L526

**Comment by scrcro** [comment]
This works somewhat better. I tracked down the loss reset issue to a bug. But even after that, I do not see it converging. It goes down by about 20% and then oscillates. I tried with step size 0.001, which is a bit better.
Will try with even smaller step size.

**Comment by scrcro** [comment]
Is this correct:
https://github.com/mlpack/mlpack/blob/e723e470684d6ebe907bababf0a9b42ac7c2a425/src/mlpack/core/optimizers/sgd/sgd_impl.hpp#L127
Or should it be incremented by 1?

**Comment by scrcro** [comment]
@zoq 
How is the `maxIteration` parameter in `SGD` supposed to be interpreted? Right now with number of samples = 300 and batchSize = 30 if I use maxIteration = 10 iterations then the whole thing is over after a single iteration. Is the `maxIteration` supposed to be some multiple of the epochs? 
The code here is unclear to me:
https://github.com/mlpack/mlpack/blob/e723e470684d6ebe907bababf0a9b42ac7c2a425/src/mlpack/core/optimizers/sgd/sgd_impl.hpp#L127-L128

**Comment by zoq** [comment]
> How is the maxIteration parameter in SGD supposed to be interpreted?
in case of SGD this equals to the number of samples visted.

``i += effectiveBatchSize;``

tracks the overall number of iteration

``currentFunction += effectiveBatchSize;``

tracks the number of functions visted; we check the performance or shuffel the data after every ``numFunctions`` (``numFunctions`` = number of samples). Perhaps it makes sense to increase ``i`` by one instead of ``effectiveBatchSize``, @rcurtin what do you think?

**Comment by rcurtin** [comment]
All of the other neural network toolkits that I know of specifically allow the user to specify the maximum number of iterations in terms of epochs (with epochs = maxIterations / numFunctions, where in the neural network case numFunctions is the number of points in the dataset).

I don't have any particular issue changing all the SGD-like optimizers to accept `epochs` instead of `maxIterations`---what do you think?  Alternate ideas would be allowing either to be specified, or changing nothing and adding a bunch of documentation.  I think as far as use cases are concerned, most of our users are more likely to want to set the number of epochs directly instead of needing to do something like `sgd.MaxIterations() = 10 * data.n_cols` or something.

**Comment by rcurtin** [comment]
I also opened #1481 to continue the discussion/prototyping of the `FunctionWrapper` idea we were talking about earlier.

**Comment by scrcro** [comment]
Most frameworks/tools treats the number of epochs as a parameter and the iterations are dependent on the number of training samples. Therefore using the number of epochs as the parameter would be most easily understood.

**Comment by rcurtin** [comment]
I think that the primary part of the work here is discussed in #1481 now, so just in the interest of keeping the bug tracker neat I'll close this one.
